src/storage/AdminProcessor.h

```diff
@@ -125,8 +200,54 @@ class WaitingForCatchUpDataProcessor : public BaseProcessor<cpp2::AdminExecResp>
         return new WaitingForCatchUpDataProcessor(kvstore);
     }
 
+    ~WaitingForCatchUpDataProcessor() {
+    }
+
     void process(const cpp2::CatchUpDataReq& req) {
-        UNUSED(req);
+        auto spaceId = req.get_space_id();
+        auto partId = req.get_part_id();
+        LOG(INFO) << "Received waiting for catching up data for space "
+                  << spaceId << ", part " << partId;
+        auto ret = kvstore_->part(spaceId, partId);
+        if (!ok(ret)) {
+            onFinished(to(error(ret)));
+            return;
+        }
+        auto part = nebula::value(ret);
+        auto peer = kvstore::NebulaStore::getRaftAddr(HostAddr(req.get_target().get_ip(),
+                                                               req.get_target().get_port()));
+
+        folly::async([this, part, peer, spaceId, partId] {
+            int retry = FLAGS_waiting_catch_up_retry_times;
+            while (retry-- > 0) {
+                LOG(INFO) << "Waiting for catching up data, peer " << peer
+                          << ", try " << retry << " times";
+                auto res = part->isCatchedUp(peer);
+                switch (res) {
+                    case raftex::AppendLogResult::SUCCEEDED:
+                        onFinished(cpp2::ErrorCode::SUCCEEDED);
+                        return;
+                    case raftex::AppendLogResult::E_INVALID_PEER:
+                        onFinished(cpp2::ErrorCode::E_INVALID_PEER);
+                        return;
+                    case raftex::AppendLogResult::E_NOT_A_LEADER: {
+                        auto leaderRet = kvstore_->partLeader(spaceId, partId);
+                        CHECK(ok(leaderRet));
+                        auto leader = value(std::move(leaderRet));
+                        resp_.set_leader(toThriftHost(leader));
+                        onFinished(cpp2::ErrorCode::E_LEADER_CHANGED);
+                        return;
+                    }
+                    case raftex::AppendLogResult::E_SENDING_SNAPSHOT:
+                        LOG(INFO) << "Still sending snapshot, please wait...";
+                        break;
+                    default:
+                        LOG(INFO) << "Unknown error " << static_cast<int32_t>(res);
+                        break;
+                }
+                sleep(FLAGS_waiting_catch_up_interval_in_secs);
+            }
```

Comment:
Maybe need to return resp if retry times exceeds limit?

Reply:
Good catch